### ISSUE TITLE
False alarm for `<dyn Fn(&T) -> String as FnOnce<(&T)>>::Output` doesn't implement `Display`

### ISSUE THREAD

**Issue by songzhi** [intellij-rust/intellij-rust#5817]
## Environment

* **IntelliJ Rust plugin version:** 0.3.127.3249-201
* **Rust toolchain version:** 1.47.0-nightly (d6953df14 2020-07-25) x86_64-pc-windows-msvc
* **IDE name and version:**  CLion 2020.2 RC
* **Operating system:** Windows10 19041 

## Problem description
False alarm like this, no warning or error with `Clippy` and run successfully.

![image](https://user-images.githubusercontent.com/18499849/88633129-28dfcd80-d0e7-11ea-9ffe-8e546238649f.png)

## Steps to reproduce

```rust
type Serializer<T> = dyn Fn(&T) -> String;

struct Movie {
    title: String,
    year: isize,
}

impl Movie {
    pub fn print(&self, serialize: &Serializer<Self>) {
        println!("{}", serialize(self));
    }
}

fn movie_to_json(movie: &Movie) -> String {
    format!("{} in {}", movie.title, movie.year)
}

fn main() {
    Movie {
        title: "The Room".into(),
        year: 2003,
    }
    .print(&movie_to_json);
}

```

**Comment by thomasweitzel** [comment]
This bug pops up everytime I write a curried function and do partial application. Like so:
```rust
fn f(a: i32) -> Box<dyn Fn(i32) -> i32> {
    Box::new(move |b| a + b)
}
// ...
let g = f(10);
println!("{}", g(32)); // Will print 42
```
Makes functional programming in Jetbrains IDEs a pain.

**Comment by Kobzol** [comment]
@vlad20012 I tried to dig into this and it seems that the problem is in type inference of the `Output` associated type of `Fn`-like traits. The type in question is `&dyn Fn(&Movie) -> String`, for which `ImplLookup::lookupAssociatedType` returns `null`. The type is being attempted to be found using `lookupAssocTypeInSelection`, but the impl block looks like this:
```rust
#[stable(feature = "rust1", since = "1.0.0")]
impl<A, F: ?Sized> Fn<A> for &F
where
    F: Fn<A>,
{
    extern "rust-call" fn call(&self, args: A) -> F::Output {
        (**self).call(args)
    }
}
```
As there is not explicit `Output` associated type specified, the corresponding type is not found.

It works when `FnOnce` behind a `Box`, then the corresponding type is found via `lookupAssocTypeInBounds`.

**Comment by vlad20012** [comment]
Oh. We really should looks `Output` type in `FnOnce` impl. There are too many hacks around function traits...
